base/common/src/main/java/org/dogtagpki/client/NonBlockingSocketFactory.java

```diff
@@ -0,0 +1,189 @@
+            HttpParams params)
+            throws IOException,
+            UnknownHostException {
+
+        String hostname = null;
+        int port = 0;
+        if (remoteAddress != null) {
+            hostname = remoteAddress.getHostName();
+            port = remoteAddress.getPort();
+        }
+
+        int localPort = 0;
+        InetAddress localAddr = null;
+
+        if (localAddress != null) {
+            localPort = localAddress.getPort();
+            localAddr = localAddress.getAddress();
+        }
+
+        SSLSocketFactory socketFactory;
+        try {
+            CryptoManager.getInstance();
+
+            KeyManagerFactory kmf = KeyManagerFactory.getInstance("NssX509", "Mozilla-JSS");
+            KeyManager[] kms = kmf.getKeyManagers();
+
+            // Create JSSTrustManager since the default JSSNativeTrustManager
+            // does not support hostname validation and cert approval callback.
+            //
+            // JSSTrustManager currently does not support cert validation with
+            // OCSP and CRL.
+            //
+            // TODO: Fix JSSTrustManager to support OCSP and CRL, then replace
+            // DefaultSocketFactory with this class.
+
+            JSSTrustManager trustManager = new JSSTrustManager();
+            trustManager.setHostname(hostname);
+            trustManager.setCallback(connection.getCallback());
+
+            TrustManager[] tms = new TrustManager[] { trustManager };
+
+            SSLContext ctx = SSLContext.getInstance("TLS", "Mozilla-JSS");
+            ctx.init(kms, tms, null);
+
+            socketFactory = ctx.getSocketFactory();
+
+        } catch (Exception e) {
+            throw new IOException("Unable to create SSL socket factory: " + e.getMessage(), e);
+        }
+
+        JSSSocket jssSocket;
+        try {
+            if (socket == null) {
+                logger.info("Creating new SSL socket");
+                jssSocket = (JSSSocket) socketFactory.createSocket(
+                        InetAddress.getByName(hostname),
+                        port,
+                        localAddr,
+                        localPort);
+
+            } else {
+                logger.info("Creating SSL socket with existing socket");
+                jssSocket = (JSSSocket) socketFactory.createSocket(
+                        socket,
+                        hostname,
+                        port,
+                        true);
+            }
+
+        } catch (Exception e) {
+            throw new IOException("Unable to create SSL socket: " + e.getMessage(), e);
+        }
+
+        jssSocket.setUseClientMode(true);
+
+        String certNickname = connection.getConfig().getCertNickname();
+        if (certNickname != null) {
+            logger.info("Client certificate: "+certNickname);
+            jssSocket.setCertFromAlias(certNickname);
+        }
+
+        jssSocket.getEngine().setListeners(Arrays.asList(new SSLSocketListener() {
```

Comment:
`getEngine()` is not need, possible to call [setListeners()](https://github.com/dogtagpki/jss/blob/master/base/src/main/java/org/mozilla/jss/ssl/javax/JSSSocket.java#L458)

Reply:
Fixed.